scripts/snapshot-calculator.ts

```diff
@@ -0,0 +1,161 @@
+    stakeLogs.map(async (currentLog) => {
+      if (currentLog.blockNumber <= blockNumber) {
+        //Removing the padding zeros
+        const currentAddress = utils.hexStripZeros(currentLog.topics[1]);
+        if (!stakers.includes(currentAddress)) {
+          stakers.push(currentAddress);
+        }
+      }
+    }),
+  );
+  return stakers;
+};
+
+const getStakers = async (
+  blockNumber: number,
+  minStakedAmount: number,
+  stakingPoolAddress: string,
+  provider: JsonRpcProvider,
+) => {
+  stakingContract = (await ethers.getContractFactory("StakingPoolPatronKYC", wallet.connect(provider))).attach(
+    stakingPoolAddress,
+  );
+  const stakingLogs = await provider.getLogs({
+    address: stakingPoolAddress,
+    topics: ["0x270d6dd254edd1d985c81cf7861b8f28fb06b6d719df04d90464034d43412440"], // Ref of topic corresponding to `StakeAdded` event
+    fromBlock: 0,
+    toBlock: blockNumber,
+  });
+
+  const stakersList = await parseEvents(stakingLogs, blockNumber);
+
+  return stakersList;
+};
+
+const calculateSnapshot = async (
+  _stakers: string[],
+  chainID: number,
+  rpcProvider: providers.JsonRpcProvider,
+  blockNumber: number,
+  minBalance: number,
+  _stakingAddress?: string,
+) => {
+  const _stakingContract = _stakingAddress ? _stakingAddress : process.env.STAKINGPOOL;
+  const snapshots: Snapshot[] = [];
+
+  await Promise.all(
+    _stakers.map(async (currentAddress) => {
+      let stakingAmount = null;
+      try {
+        do {
+          stakingAmount = await _rpcReadContractSlot(
+            STAKES_STORAGE_SLOT,
+            currentAddress,
+            rpcProvider,
+            blockNumber,
+            _stakingContract,
+          );
+          if (stakingAmount === null) {
+            nodeEmoji.emojify(":x:", "RPC connection broke ... Recalculate ", currentAddress);
+          } else {
+            console.log(nodeEmoji.emojify(":white_check_mark:"), currentAddress);
+          }
+        } while (stakingAmount === null);
+
+        if (Number(stakingAmount) >= minBalance) {
+          snapshots.push({
+            did: formatDID(currentAddress, chainID),
+            issuerFields: [
+              {
+                stakeAmount: Number(stakingAmount),
+                minimumBalance: minBalance,
+                snapshotBlock: blockNumber,
+                chainId: chainID,
+                stakingPoolAddress: _stakingContract as string,
+              },
+            ],
+          });
+          //We remove this log from failingCalculation array if it is present
+          if (failingCalculations.includes(currentAddress)) {
+            failingCalculations = failingCalculations.filter((addr) => {
+              return addr !== currentAddress;
+            });
+          }
+        }
+        return snapshots;
+      } catch (error) {
+        console.log(`An error occurred during calculation of snapshot ${currentAddress} ... `);
+        // We keep track of all transactions which fail
```

Comment:
This comment is a bit misleading IMO as there aren't transactions as not blockchain state is changing. Would you agree with this @knzeng-e ?
```suggestion
        // We keep track of all snapshot calculations which fail
```

Reply:
oh yes, I agree @jrhender !